Models

cbrom opened the pull request:
- [x] models finalized
- [x] model relationships implemented
- [x] sample r/ship functions created

cbrom committed:
course model implemented

cbrom committed:
answer options model implemented

cbrom committed:
basic certifate model implemented

cbrom committed:
user id added to certificates

cbrom committed:
basic course author manytomany relationship created

cbrom committed:
basic evaluation criteria model implemented

cbrom committed:
basic level model implemented

cbrom committed:
basic quiz questions model added

cbrom committed:
basic user answers for quiz implemented

cbrom committed:
basic quiz model created

cbrom committed:
basic content block model implemented

cbrom committed:
basic target model implemented

cbrom committed:
basic target version model implemented

cbrom committed:
basic target groups model implemented

cbrom committed:
certificate changed to be a basic certificate and not a user issued certificate

cbrom committed:
basic model for issued certificates implemented

cbrom committed:
type (Quize)

cbrom committed:
remove citext tag, as its not supported

cbrom committed:
models migration modified to include new ones

cbrom committed:
basic common functions moved to base struct

cbrom committed:
answer options crud functions implemented

cbrom committed:
certificate crud functions implemented

cbrom committed:
conent block crud functions implemented

cbrom committed:
course author crud functions implemented

cbrom committed:
course crud functions implemented

cbrom committed:
evaluation criteria crud functions implemented

cbrom committed:
issued certificates crud functions implemented

cbrom committed:
level crud functions implemented

cbrom committed:
quiz related crud functions implemented

cbrom committed:
target and target group related crud functions implemented

cbrom committed:
header logo added

cbrom committed:
Merge branch 'dev' into models

cbrom committed:
students and models table created (many to many)

cbrom committed:
students course model migrated

cbrom committed:
crud method pointer names changed to match model name
- belongs to relationships created
- has many relationships created

cbrom committed:
answer GetQuestions function implemented

cbrom committed:
quiz answer and answer options r/ship functions implemented

Azrakal commented on cmd/models/answer_option.model.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+// AnswerOptionList defines array of answer options objects
+type AnswerOptionList []*AnswerOption
```
How come you're defining this list type everywhere?

cbrom replied:
Eventually, I am going to need to fetch a list of records, so instead of creating the lists on the fly, I figured it would be convenient to define it once

Azrakal replied:
No need to do this IMO. Go already has a concise slice syntax.

cbrom replied:
What do you mean? I am just defining a type beforehand since I am going to use it many times. I could always pass around []*Model on the fly... but with a defined type I can add more functions to it like filtering ...

Azrakal replied:
I just don't think you gain anything by having a type definition here. You can always write functions for `[]*AnswerOption` just like you can write functions for `AnswerOptionList`. 

The thing is that you're not doing anything with this type definition. Usually if you want to define a type alias (i.e., not a struct definition), you want to limit the scope of the accepted values for a function. For example, I could do something like this:

```
type IceCream string

const (
    Vanilla IceCream = "Vanilla"
    Chocolate IceCream = "Chocolate"
)

func GiveIceCream() IceCream {...}
```

to limit the strings returned by my function. But defining type aliases just like `type IceCream string` doesn't really do anything in my opinion.

Just my two cents, ultimately it's not a huge deal. Just doesn't really seem idomatic to me.

cbrom replied:
I see. I will remove them.

Azrakal replied:
One other thing you can use type definitions to do is to formalize general data structures. For example, I could define `type ParameterMap map[string]string` if I'm using this specific concept in multiple places.

cbrom replied:
yeah, actually that's why I defined the List types

Azrakal replied:
The difference is that `map[string]string` is general. If I'm defining `type ParameterMap map[string]string`, I'm saying "this is a map of my Parameters". 

IMO `type AnswerOptionList []*AnswerOption` doesn't add any extra information.

Again, this is stylistic, so it's up to you if you want to keep it!

Azrakal commented on cmd/models/certificate.model.go:
```diff
@@ -0,0 +1,84 @@
+
+	Course             *Course               `gorm:"foreignkey:CourseID"`
+	Issuer             *CourseAuthor         `gorm:"foreignkey:IssuerID"`
+	IssuedCertificates IssuedCertificateList `gorm:"foreignkey:CertificateID"`
```
Commenting for here and everywhere else you have these foreign key relations. 

Have you tested that this works? Is gorm automatically doing the join for you here? I've never seen the DB structs written this way.

cbrom replied:
I have not started testing (yet). But I reckon that the tags will help gorm build a relationship when performing ORM actions like `Related` function.

This does not create actual relationship in the DB tho.

Azrakal replied:
I would definitely try writing some tests to make sure this works first. You may find that this doesn't work.

cbrom replied:
alright, I will do that.

Azrakal reviewed (commented):
A bit skeptical without tests that these models will actually work. Can you add some database tests with a local DB?

cbrom committed:
user model handler not set exception added

cbrom committed:
handler not set error added

cbrom committed:
ginkgo added

cbrom committed:
ginkgo test orchastrator created

cbrom committed:
creating a new user test added

cbrom committed:
handle error for email duplicate in user registration test

cbrom committed:
TEMPORARY testing environment for models created

cbrom committed:
user model soft delete and permanent delete separated.

cbrom committed:
softdelete function implemented and permanent delete identified

cbrom committed:
close db function created

cbrom committed:
restructuring of the test

cbrom committed:
data generator test util created

cbrom committed:
basic CRUD tests performed

cbrom committed:
move suite related configuration to models suite

cbrom committed:
create new course author function implemented

cbrom committed:
course author (basic CRUD) tests performed

cbrom committed:
separate soft delete and permanent delete

cbrom committed:
*Course and *User pointers changed to Course and User values, get course and get author methods created

cbrom committed:
courseauthor relation ship with author and course tests performed

cbrom committed:
get authored coures method implemented

cbrom committed:
create a student course mock function created

cbrom committed:
get student courses method added to users model

cbrom committed:
user get courses method tested

cbrom committed:
create level mock function implemented

cbrom committed:
move test functions to "models_test" package and reflet changes on the tests that user model functions

cbrom committed:
answer option get Question method fixed to fetch question properly

cbrom committed:
replacement of test.utils.go

cbrom committed:
errHandlerNotSet error var added

cbrom committed:
user handlerNotSet changed to errHandlerNotSet

cbrom committed:
models create function moved from tests to util_test

cbrom committed:
softdelet and permanent delete functions separated
get course method implemented

cbrom committed:
basic level crud tests and course relationship tests implemented

cbrom committed:
get target groups method implemented

cbrom committed:
create target function implemented

cbrom committed:
level target group relationships tested

cbrom committed:
get certificates of course author method implemented

cbrom committed:
create new certificate mock function implemented

cbrom committed:
- softdelete and permanent delete separated
- get Issuer and get course methods implemented
- issuer changed to courseauthor as it causes an invalid association issues

cbrom committed:
issuer changed to courseauthor

cbrom committed:
TargetGroupID changed to LevelID -- but this did not cause any problems in the associations

cbrom committed:
change issuer id to author id

cbrom committed:
course author certificate relationship tested

cbrom committed:
certificate tests
crud
   - fetch
   - update
   - softdelete

relationship
   - get course
   - get issuer

cbrom committed:
student course tests
functions
- get course
- get user
- soft delete
- delete

cbrom committed:
student course tests
- crud
   - fetch
   - update
   - delete
- relationship
   - get coures
   - get user

cbrom committed:
fix create student method to create the studentcourse record instead of user

cbrom committed:
course relationship functions implemented

cbrom committed:
change comment (from copy paste)

cbrom committed:
relationship tests added
-level
-author
-student
-certificate

cbrom committed:
target improvement
 * target group improvement
 - get target method added
 - get targets method added
 - delete and soft delete functions separated

cbrom committed:
target model improvements
- get target group method implemented
- get target versions method impelemented
- get quizzes method implemented
- delete and softdelete methods separated

cbrom committed:
create target function for test added

cbrom committed:
create new target version mock added

cbrom committed:
target group tests
CRUD
- fetch
- create
- update
- delete
Relationship
- get level
- get targets

cbrom committed:
create quiz mock added

cbrom committed:
target version improvements
- get target method implemented
- get content blocks method implemented

- delete and soft delete functions separated

cbrom committed:
target model tests
CRUD
- create
- fetch
- update
- delete

relationship
- target group
- target versions
- quiz

cbrom committed:
quiz model improvements
- get target method implemented
- get quiz questions method implemented

- soft delete and delete functions separated

cbrom committed:
quiz model tests
CRUD
- create
- update
- delete
- fetch
Relationsihip
- target based
- quiz questions based

cbrom committed:
ginkgo

cbrom committed:
change Correct answer id from uuid to just string as it is causing problem

cbrom committed:
separate delete and soft delete

cbrom committed:
remove answeroption ao as it is unncessary

cbrom committed:
minor changes

cbrom committed:
create answer option methiod implemented

cbrom committed:
quiz question model tests
CRUD
- create
- update
- fetch
- delete
Relationships
- quiz
- answers
- quiz answer (correct answer)

cbrom committed:
content block improvement
- create GetTargetVersion method
- separate delete and softdelete

cbrom committed:
target version model tests
CRUD
- create
- update
- fetch
- delete
Relationship
- get target
- get content blocks

cbrom committed:
CreateContentBlock function mock implemented

cbrom committed:
get correct answer method updated to select with query rather than relationship

cbrom committed:
quiz question model test updated according to change

cbrom committed:
answer option model tests
CRUD
- create
- update
- fetch
- delete
Relationship
- get quiz question

cbrom committed:
remove some comments

cbrom committed:
content block model tests
CRUD
- create
- update
- delete
Relationship
- get quiz method
* (get target version)
\

cbrom committed:
issued certificate improvements
- get certificate method
- separate delete and soft delete

cbrom committed:
comment

cbrom committed:
create issued certificate mock

cbrom committed:
issued certificate model tests

cbrom committed:
test database hard coded for safety

cbrom committed:
git actions cmd/model test included

cbrom committed:
fix models test actions

cbrom committed:
fix models test actions

cbrom committed:
fix models test actions

cbrom committed:
removed model tests from git actions

cbrom committed:
evaluation criteria improvements
- get course method implemented
- delete and soft delete methods separated

cbrom committed:
create evaluation criteria mock function implemented

cbrom committed:
evaluation criteria tests added

cbrom committed:
evaluation criteria relattionship test added to course

Azrakal commented on cmd/models/content_block.model_test.go:
```diff
@@ -0,0 +1,113 @@
+	})
+
+	AfterEach(func() {
+		u := models.User{}
+		u.Delete()
+
+		c := models.Course{}
+		c.Delete()
+
+		l := models.Level{}
+		l.Delete()
+
+		tg := models.TargetGroup{}
+		tg.Delete()
+
+		t := models.Target{}
+		t.Delete()
+
+		tv := models.TargetVersion{}
+		tv.Delete()
+
+		cb := models.ContentBlock{}
+		cb.Delete()
```
Small comment: you can probably factor this out into a utility method, since you're doing it in quite a few places.

Azrakal commented on cmd/models/course.model_test.go:
```diff
@@ -0,0 +1,141 @@
+		c.Delete()
+	})
+
+	Describe("Basic Crud Tests", func() {
+		Context("CRUD basics", func() {
```
Should be 'Context' and 'When' here instead of 'Describe' and 'Context'?

Azrakal reviewed (approved):
I took a look at your tests and they seem to be set up pretty well. I think some of the BDD descriptions could use some work, but I wouldn't spend too much time on it.

Another thing I'd do is just merge this now and open another branch because it's getting pretty massive!
